src/aapns/connection.py

```diff
@@ -0,0 +1,476 @@
+        ):
+            raise ValueError("Origin must be https://<host>[:<port>]")
+
+        host = url.hostname
+        port = url.port or 443
+
+        ssl_context = ssl if ssl else create_ssl_context()
+        if (
+            OP_NO_TLSv1 not in ssl_context.options  # type: ignore # https://github.com/python/typeshed/issues/3920
+            or OP_NO_TLSv1_1 not in ssl_context.options  # type: ignore
+        ):
+            raise ValueError("SSL Context cannot allow TLS 1.0 or 1.1")
+
+        # https://bugs.python.org/issue40111 validate context h2 alpn
+
+        protocol = h2.connection.H2Connection(
+            h2.config.H2Configuration(client_side=True, header_encoding="utf-8")
+        )
+        protocol.local_settings = h2.settings.Settings(
+            client=True,
+            initial_values={
+                # Apple server settings:
+                # HEADER_TABLE_SIZE 4096
+                # MAX_CONCURRENT_STREAMS 1000
+                # INITIAL_WINDOW_SIZE 65535
+                # MAX_FRAME_SIZE 16384
+                # MAX_HEADER_LIST_SIZE 8000
+                h2.settings.SettingCodes.ENABLE_PUSH: 0,
+                h2.settings.SettingCodes.MAX_CONCURRENT_STREAMS: 2 ** 20,
+                h2.settings.SettingCodes.MAX_HEADER_LIST_SIZE: 2 ** 16 - 1,
+                h2.settings.SettingCodes.INITIAL_WINDOW_SIZE: MAX_RESPONSE_SIZE,
+            },
+        )
+
+        protocol.initiate_connection()
+        protocol.increment_flow_control_window(CONNECTION_WINDOW_SIZE)
+
+        read_stream, write_stream = await wait_for(
+            open_connection(
+                host, port, ssl=ssl_context, ssl_handshake_timeout=TLS_TIMEOUT
+            ),
+            CONNECTION_TIMEOUT,
+        )
+        try:
+            info = write_stream.get_extra_info("ssl_object")
+            if not info:
+                raise Closed("Failed TLS handshake")
+            proto = info.selected_alpn_protocol()
+            if proto != "h2":
+                raise Closed("Failed to negotiate HTTP/2")
+        except Closed:
+            write_stream.close()
+            with suppress(SSLError, ConnectionError):
+                await write_stream.wait_closed()
+            raise
+
+        # FIXME we could wait for settings frame from the server,
+        # to tell us how much we can actually send, as initial window is small
+        return cls(host, port, protocol, read_stream, write_stream)
+
+    def __post_init__(self):
+        self.should_write = asyncio.Event()
+        self.should_write.set()
+        self.reader = create_task(self.background_read(), name="bg-read")
+        self.writer = create_task(self.background_write(), name="bg-write")
+
+    async def post(self, request: "Request") -> "Response":
+        """Post the `request` on the connection"""
+        if len(request.body) > MAX_NOTIFICATION_PAYLOAD_SIZE:
+            raise ValueError("Request body is too large")
+
+        request.time_left_or_time_out
```

Comment:
this looks like a bug in the code (you're accessing a property without doing anything with it). This property should probably be a method.

Reply:
yeah... it's like... I dunno how to make this API nice.

It's both a property and a side effect:
I've moved raising timeout to the request obj, because I figured I want to give more information about the timeout

So... I dunno what would be the best really.

Reply:
`async def get_time_left()` seems reasonable to me. maybe `get_time_left_or_fail()` to indicate erroring is a normal thing for this method.

Reply:
👍